assignment/final/aLife/code/master.js

```diff
@@ -0,0 +1,417 @@
+let pkmnPic = document.getElementById("terminalScreen");
+let pkmnStats = document.getElementById("statSection");
+let moveset = document.getElementById("scrollingMoveset");
+let type1 = document.getElementById("leftType").getElementsByTagName("p")[0];
+let type2 = document.getElementById("rightType").getElementsByTagName("p")[0];
+let idBox = document.getElementById("pkmnIdBox");
+let downId = document.getElementById("downId");
+let upId = document.getElementById("upId");
+let pkmnName = document.getElementById("pkmnName");
+let terminalButton = document.getElementById("terminalButton");
+let typesDiv = document.getElementById("types");
+let clearFilter = document.getElementById("clearFilter");
+let selectedType1;
+let selectedType2;
+
+pkmnListDiv.innerHTML = "Loading Pokemon";
+
+function fetchPkmn() {
+    fetch(baseUrl + 'pokedex/1/')
+        .then(res => res.json())
+        .then(data => {
+            for (let i = 0; i < data.pokemon_entries.length; i++) {
+                pkmnList.push({
+                    ID: data.pokemon_entries[i].entry_number,
+                    NAME: data.pokemon_entries[i].pokemon_species.name,
+                    PKMNURL: data.pokemon_entries[i].pokemon_species.url
+                });
+            }
+            pkmnListDiv.innerHTML = "";
+            for (let i = 0; i < pkmnList.length; i++) {
+                pkmnListDiv.innerHTML += `<p id="${pkmnList[i].NAME}">${pkmnList[i].NAME}</p>`;
+            }
+            //console.log(pkmnList);
+        });
+}
+
+function fetchTypes() {
+    fetch(baseUrl + 'type/')
+        .then(res => res.json())
+        .then(data => {
+            for (let i = 0; i < data.results.length; i++) {
+                pkmnTypes.push({
+                    TYPE: data.results[i].name,
+                    TYPEURL: data.results[i].url
+                });
+            }
+            //console.log(pkmnTypes);
+            for (let i = 0; i < pkmnTypes.length; i++) {
+                typesCorespondingPkmn.push({
+                    TYPE: pkmnTypes[i],
+                    LIST: []
+                });
+
+                let urlForList = pkmnTypes[i].TYPEURL;
+                fetch(urlForList)
+                    .then(res => res.json())
+                    .then(data => {
+                        //console.log(data.pokemon);
+                        typesCorespondingPkmn[i].LIST = data.pokemon;
+                    });
+            }
+            //console.log(typesCorespondingPkmn);
+        });
+    //console.log(moves);
+    //console.log(pkmnTypes);
+}
+
+function fetchMoves() {
+    fetch(baseUrl + 'move/')
+        .then(res => res.json())
+        .then(data => {
+            let num = data.count;
+            fetch(baseUrl + 'move?offset=0&limit=' + num)
+                .then(res => res.json())
+                .then(data => {
+                    for (let i = 0; i < num; i++) {
+                        fetch(data.results[i].url)
```

Comment:
If you ever read an article about JavaScript and they mention the staircase to hell...this is what they mean.  How `asynchronous` calls keep getting nested further in further down in a staircase manner.  You can do a few things to solve this

1) return back promises
```
fetch('../url/1')
 .then((response1) => {
  return fetch('../url/2')
 })
 .then((response2) => {
  return fetch('../url/3')
 })
.catch( (err) => {
// If any errors occur...this catches for all
console.error(err)
})
```

2) `async` function (which really build out `Promise` like objects).  This method will give you a synchronous code feeling...but actually be doing asynchronous work.

```
async function fetchMoves() {
 const {data: moveData} = await fetch('../url/1')
 const {data: additionalMoves } = await fetch('../url2')
...etc.,
}
```